hive-metastore/src/main/java/org/apache/iceberg/hive/MetastoreUtil.java

```diff
@@ -48,14 +50,28 @@ public class MetastoreUtil {
   private MetastoreUtil() {}
 
   /**
-   * Calls alter_table method using the metastore client. If possible, an environmental context will
-   * be used that turns off stats updates to avoid recursive listing.
+   * Calls alter_table method using the metastore client. If the HMS supports then, environmental
+   * context with will be set in a way that turns off stats updates to avoid recursive file listing.
    */
   public static void alterTable(
       IMetaStoreClient client, String databaseName, String tblName, Table table) {
-    EnvironmentContext envContext =
-        new EnvironmentContext(
-            ImmutableMap.of(StatsSetupConst.DO_NOT_UPDATE_STATS, StatsSetupConst.TRUE));
-    ALTER_TABLE.invoke(client, databaseName, tblName, table, envContext);
+    alterTable(client, databaseName, tblName, table, ImmutableMap.of());
+  }
+
+  /**
+   * Calls alter_table method using the metastore client. If the HMS supports then, environmental
+   * context with will be set in a way that turns off stats updates to avoid recursive file listing.
+   */
+  public static void alterTable(
+      IMetaStoreClient client,
+      String databaseName,
+      String tblName,
+      Table table,
+      Map<String, String> extraEnv) {
+    Map<String, String> env = Maps.newHashMapWithExpectedSize(extraEnv.size() + 1);
+    env.putAll(extraEnv);
+    env.put(StatsSetupConst.DO_NOT_UPDATE_STATS, StatsSetupConst.TRUE);
+
```

Comment:
Maybe to validate hive version, one way is  check here if alter_table method has enough arguments for the env?  If not, I feel the whole scheme is not going to work.

Reply:
Sadly no 😢 . Historical reasons again

Reply:
Yea I was thinking, can we throw a ValidationException here then, if the API is to pass in env context, and the alter_table method that we load does not take it?

Reply:
We can not distinguish between the 2 cases. The API is backwards compatible (as expected by the Hive team), so the HMS will always take the env context input variables, but will not act on them.

I think this is not that interesting case, as the writer should have a clear info about the HMS they are connecting to.

Reply:
Oh maybe I was not clear, I meant the alter_method that we load dynamically:  https://github.com/apache/iceberg/blob/master/hive-metastore/src/main/java/org/apache/iceberg/hive/MetastoreUtil.java#L29  .  Original question was just asking if our method object is the one loaded without env, and we pass in env here, what will happen. (is error message good enough)

Reply:
The "env"-less alter table will only be used with old HMS clients so we do not have issues with adding env variables to the HMS call which does not handle it.

We still might provide env values which the server does not act upon (I was talking about this case in my previous comments here), but at least we can be sure that we do not break the API calls

Reply:
I was looking through DynMethod, there is no check and will just pass whatever args it can fit.  I think it will be nice to add a 'argLength' method DynMethod and then assert that argLength == 5 here, so ensure the env is not silently dropped.

Reply:
This method is used in other cases as well and doing some extra checks would be edgy.

Added a check for the `NoLock` constructor to ensure that at least Hive 2 client is used when a `NoLock` object is created. This ensures that the correct client is used in this case.

Reply:
Hey sorry , I thought I commented on this, but must have missed it.  I was thinking something like:

if (env.size() > 0) {
  Preconditions.checkArgument("Environment context is non-empty but alter_table method does not support it",  ALTER_TABLE.args().size() == 5);
}

What do you think?  Isn't alter_table only called with envContext is the new code?

Reply:
We always add `ImmutableMap.of(StatsSetupConst.DO_NOT_UPDATE_STATS, StatsSetupConst.TRUE)` which is an optimisation for skipping stats generation / file listing when it is not needed on alter table commands. This always depended on DynMethod skipping last parameters (maybe by accident)

See: https://github.com/apache/iceberg/blob/fede493d59f17ff2bfc0744b296d90bd36130386/hive-metastore/src/main/java/org/apache/iceberg/hive/MetastoreUtil.java#L56-L59

We can modify the proposed command to:
```
if (env.size() > 1) {
  Preconditions.checkArgument("Environment context is non-empty but alter_table method does not support it", 
  ALTER_TABLE.args().size() == 5);
}
```

Or add a new check to the the new `alterTable` method, but I find it better / more readable to add the check to the NoLock constructor.

WDYT?

Reply:
Yea I was thinking at the very beginning of the method, before we put the ALTER_TABLE?  Yea it may be better for understanding of the code if we have both checks, but you are right it's not a big deal in practice, just in the rare case there's a patched version of Hive 2 out there that has alter_table without 5 arguments.

Reply:
Checked, and every Hive 2 should have the required method on the HMS API:
https://github.com/apache/hive/blob/release-2.0.0/metastore/if/hive_metastore.thrift

So we are good here